### PR TITLE
Add V1 Lease Controller

### PR DESCRIPTION
**Pull request by sargun**

### CODE REVIEW

File: internal/leasecontroller/lease_controller.go [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+type LeaseController interface {
+	// leaseControllerStatus returns whether the Lease Controller has successfully been able to establish a lease.
+	LeaseControllerStatus(ctx context.Context) Status
+	// Run starts the lease controller. It blocks until context is cancelled, or if the lease cannot be established.
+	Run(ctx context.Context)
+}
+
+// NewV1BetaV1LeaseController creates a new lease controller.
+//
+// shouldUpdateOrEstablishLeaseFunc is a function that may block, but will be called every lease update interval,
```
Comment (pires) [inline internal/leasecontroller/lease_controller.go:40]: `shouldUpdateOrEstablishLeaseFunc` doesn't exist (anymore?) and it's actually `Run`.

Also, my understanding of the code below is that the `Run` function is indeed blocking on `lc.poll(ctx)` which means this function may actually not run within the lease update interval, i.e. `lc.poll(ctx)` didn't complete within lease interval.

File: internal/leasecontroller/lease_controller.go [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+	if err != nil {
+		return fmt.Errorf("Received error when attempting to ascertain node status: %w", err)
+	}
+
+	if pr.Error != nil {
+		return newNodeNodeReady(pr)
+	}
+
+	lc.lease.Spec.RenewTime = &metav1.MicroTime{Time: time.Now()}
+	// This is 25 due to historical reasons. It was supposed to be * 5, but...reasons
```
Comment (pires) [inline internal/leasecontroller/lease_controller.go:159]: What reasons?

File: internal/nodepingcontroller/node_ping_controller.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+func (npc *nodePingController) Run(ctx context.Context) {
+	const key = "key"
+	sf := &singleflight.Group{}
+
+	// 1. If the node is "stuck" and not responding to pings, we want to set the status
+	//    to that the node provider has timed out responding to pings
+	// 2. We want it so that the context is cancelled, and whatever the node might have
+	//    been stuck on uses context so it might be unstuck
+	// 3. We want to retry pinging the node, but we do not ever want more than one
+	//    ping in flight at a time.
```
Comment (pires) [inline internal/nodepingcontroller/node_ping_controller.go:62]: This comment is concise and clear but the code to do it (cond, ticker, waiter) requires a bit of mental effort to follow, for a dumb programmer such as myself.

**Review by pires — COMMENTED** [review]
Overall it does look like it works.

File: internal/leasecontroller/lease_controller.go (hunk 2 above)
Comment (sargun) [reply internal/leasecontroller/lease_controller.go:159]: someone multiplied times 5 on input..and then on 5 again. In my limited testing in prod, suddenly changing this value doesn't seem to make other components happy.

File: internal/nodepingcontroller/node_ping_controller.go (hunk 3 above)
Comment (pires) [reply internal/nodepingcontroller/node_ping_controller.go:62]: Spent some more time w/ said code while reviewing #913. It makes more sense now. Hopefully, having that reviewed and, eventually, merged will reduce the amount of code in this PR.

**Comment by sargun** [comment]
Superceded by https://github.com/virtual-kubelet/virtual-kubelet/pull/933

**Pull request closed without merging**